rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,80 @@
+### DCP PR:
+
+# Metadata integration testing
+
+## Summary
+
+This document details an integration test design for metadata schema changes in the Human Cell Atlas (HCA) Data Coordination Platform (DCP).
+
+## Authors
+
+* [Andrey Kislyuk](mailto:akislyuk@chanzuckerberg.com)
+* [Sam Pierson](mailto:spierson@chanzuckerberg.com)
+* [Matt Weiden](mailto:mweiden@chanzuckerberg.com)
+* [Mark Diekhans](mailto:markd@ucsc.edu)
+* [Hannes Schmidt](mailto:hannes@ucsc.edu)
+
+## Shepherd
+
+[Norman Morrison](mailto:norman@ebi.ac.uk)
+
+## Motivation
+
+Currently breakages to DCP systems caused by metadata schema changes are detected at runtime upon data (bundle) upload. This forces data wranglers, DCP operators, and data consumers to react to unpredictable runtime failures rather than detecting problems before the code is merged and deployed.
+
+To alleviate this issue, the DCP development team can use a metadata schema integration test to test whether it is safe to release data downstream of the Data Storage Service (DSS), past which point applications start to depend on the schema. The benefits of this approach are that it will:
+
+1. Make the development environment more stable
+1. Make updates to software and schema smaller and more frequent
+1. Make failures due to metadata schema changes more readily observable
+1. Increase availability of DCP systems downstream of DSS
+
+## Detailed Design
+
+In this design we will use the following terms:
+ 
+ * "schema integration test system" - an automated CI/CD pipeline used to check that downstream systems can handle bundles with a new schema
+ * "new schema label" - anything that indicates that the schema has changes that have not yet been tested against downstream systems: potentially a JSON field, git tag, git branch, or comparison to tested version list
```

Comment:
I think this process is best left up to ingestion/wranglers to decide. I know @daniwelter and @malloryfreeberg had some ideas on how this would work best. What is important to this design is that it is easy to identify from the metadata itself.

From my naïve perspective, here is a way it _could_ work.

1. if there is any metadata schema changes in a bundle when you submit it to the DSS, it is marked with the new schema label
1. The new schema label could be as simple as an addition to the path string on the `describedBy` URL.

Reply:
@daniwelter @malloryfreeberg can I ping you for a comment on this? My concern is that this places yet another manual process into the list of things wranglers have to do in order to workaround a technical issue.

Reply:
Re. adapting to metadata schema changes as a technical issue: thus far, all of the strategies to protect data consumers (app developers) from changes in the metadata schema we've discussed rest on assumptions or human configuration that only a test can verify:

* Core vs. non-core schema - How do we know that core DCP components or important external data consumers (e.g. Bioconductor) do not rely on non-core schema? They do now.
* Uploader-determined major versioning - even in minor versions adding a field or changing controlled vocabulary from a field could result in broken expectations from data query results. Here's an over-simplified example: say we have all human data in the system, we start ingesting mouse, and add a "species" field. Subscribers that were previously expecting just human data would now have mouse data too (query false positive). Ultimately, metadata major/minor versioning is a human judgement which needs to be tested.
* Likely-used field blacklists/warnlists - How do we know the list is an up to date with downstream systems?
* Assertions on the schema - Client-side assertions on the schema seem like an unattractive option: all consumers (potentially third-party) of the raw metadata would have to store failed events/ids in a dead-letter queue for retry later. This is not user-friendly.

Given that none of these strategies is a sure-fire way to protect consumers of the metadata from having their expectations broken, having a test isn't a workaround to a technical issue. It's verification in the face of uncertainty.